lib/pushbox.js

```diff
@@ -0,0 +1,122 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+'use strict'
+
+const isA = require('joi')
+const Pool = require('./pool')
+const P = require('./promise')
+const {URL} = require('url')
+const validators = require('./routes/validators')
+
+const LOG_OP_RETRIEVE = 'pushbox.retrieve'
+const LOG_OP_STORE = 'pushbox.store'
+
+const SERVICE_FXA = 'fxa'
+const TOPICS_TTL = {
+  sendtab: 30 * 24 * 3600 // 30 days.
+}
+
+const PUSHBOX_RETRIEVE_SCHEMA = isA.alternatives().try(
+  // common case
+  isA.object({
+    last: isA.boolean().required(),
+    index: isA.number().required(),
+    messages: isA.array().items(isA.object({
+      index: isA.number().required(),
+      data: isA.string().regex(validators.URL_SAFE_BASE_64).required(),
+    })).optional(),
+  }),
+  // limit == 0 only returns the last index.
+  isA.object({
+    index: isA.number().required(),
+  }),
+  // error case
+  isA.object({
+    error: isA.string().required(),
+    status: isA.number().required(),
+  })
+)
+
+module.exports = function (log, config) {
+  const pool = new Pool(config.pushboxUrl, { timeout: 15000 })
+  const SafeUrl = require('./safe-url')(log)
+
+  return {
+    /**
+     * Retrieves messages for a specific device. Relays the request to Pushbox.
+     *
+     * @param {String} uid - Firefox Account uid
+     * @param {String} deviceId
+     * @param {Object} options
+     * @param {String} oAuthToken - Hex string of the user's Pushbox bearer token
+     * @param {Number} limit
+     * @param {String} [index]
+     * @returns {Promise}
+     */
+    retrieve (uid, deviceId, oAuthToken, limit, index) {
+      log.trace({
+        op: LOG_OP_RETRIEVE,
+        uid,
+        deviceId,
+        index,
+        limit
+      })
+      const query = {
+        limit: limit.toString()
+      }
+      if (index) {
+        query.index = index
+      }
+      const path = new SafeUrl('/v1/store/:service/:uid/:deviceId')
+      return pool.get(path, {service: SERVICE_FXA, uid, deviceId},
+                      {query, headers: { Authorization: `Bearer ${oAuthToken}` }})
+      .then(body => {
+        const d = P.defer()
+        PUSHBOX_RETRIEVE_SCHEMA.validate(body, (err, body) => {
```

Comment:
You could promisify `.validate` to remove the need to call `P.defer()`.

Reply:
I didn't use the "promise" API because `validate` doesn't return a real promise [0] and I ended up with test failures.

[0] https://github.com/hapijs/joi/blob/7f0c3841972d824f46fb2af295c1db91c31c6c12/lib/types/any/index.js#L676-L694